libs/fabric/src/lib/components/core/shared/changeable-helper.ts

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { EventEmitter, QueryList } from '@angular/core';
+import { Subscription } from 'rxjs';
+
+import { TypedChanges } from '../../../declarations/angular/typed-changes';
+import { ItemChangedPayload } from '../../core/declarative/item-changed.payload';
+import { ChangeableItemDirective } from './changeable-item.directive';
+
+/**
+ * Helper class for single changeable item
+ */
+export class ChangeableItemHelper<TItem> {
+  readonly onItemChanged = new EventEmitter<ItemChangedPayload<string, TItem>>();
+
+  constructor(private readonly key: string) {}
+
+  onChanges(changes: TypedChanges<TItem>) {
+    this.onItemChanged.emit({ key: this.key, changes });
+  }
+}
+
+/**
+ * Parent class for wrapper directive for multiple ChangeableItemDirectives
+ */
+export class ChangeableItemsHelper<TItem> {
+  readonly onChildItemChanged = new EventEmitter<ItemChangedPayload<string, TItem>>();
+  readonly onItemsChanged = new EventEmitter<QueryList<ChangeableItemDirective<TItem>>>();
+
+  private readonly _subscriptionsMap: { [key: string]: Subscription } = {};
+  private _changeSubscription: Subscription;
+
+  constructor(
+    private directiveItems: QueryList<ChangeableItemDirective<TItem>>,
+    private self?: IChangeableItemsContainer<TItem>,
+    private nonSelfHandler?: (nonSelfDirectives: ChangeableItemDirective<TItem>[]) => void
+  ) {}
+
+  afterContentInit() {
+    this._subscribeNewDirectives();
+    this._changeSubscription = this.directiveItems.changes.subscribe(newValues => {
+      this.onItemsChanged.emit(newValues);
+      this._subscribeNewDirectives();
+    });
+  }
+
+  onDestroy() {
+    Object.keys(this._subscriptionsMap).forEach(key => this._subscriptionsMap[key].unsubscribe());
+    this._changeSubscription.unsubscribe();
+  }
+
+  private _subscribeNewDirectives() {
+    const nonSelfDirectives = this._handleNonSelfDirectives();
+    nonSelfDirectives.forEach(directiveItem => {
+      if (this._subscriptionsMap[directiveItem.key]) {
+        this._subscriptionsMap[directiveItem.key].unsubscribe();
+      }
+      this._subscriptionsMap[directiveItem.key] = directiveItem.onItemChanged.subscribe(changes => {
+        this._handleNonSelfDirectives();
+        this.onChildItemChanged.emit(changes);
+        if (this.self && this.self.onItemChanged) {
+          this.self.onItemChanged.emit(changes);
+        }
+      });
+    });
+  }
+
+  private _handleNonSelfDirectives() {
+    const nonSelfDirectives = this.directiveItems.filter(directiveItem => directiveItem !== (this.self as any));
```

Comment:
Why the need to cast to `any`?

Reply:
VSC was giving me a warning since `directiveItem` is ChangeableItemDirective while `this.self` is IChangeableItemsContainer. The comparison should still be valid since when directives contain children of its own type, it will both extend ChangeableItemDirective and also implement IChangeableItemsContainer. The `as any` was mainly to get rid of the red underline in VSC.

![image](https://user-images.githubusercontent.com/41396986/48632097-e7d87180-e974-11e8-9ec3-64eeb4a42578.png)